src/main/kotlin/app/hashers/CommitCrawler.kt

```diff
@@ -14,137 +14,168 @@ import io.reactivex.Observable
+    fun getJGitObservable(git: Git, tail : RevCommit? = null) :
+        Observable<Pair<RevCommit, List<Pair<DiffEntry, EditList>>>> =
+        Observable.create { subscriber ->
+
+        val repo: Repository = git.repository
+        val revWalk = RevWalk(repo)
+        val head: RevCommit =
+            try { revWalk.parseCommit(repo.resolve(RepoHelper.MASTER_BRANCH)) }
+            catch(e: Exception) { throw Exception("No branch") }
+
+        val df = DiffFormatter(DisabledOutputStream.INSTANCE)
+        df.setRepository(repo)
+        df.setDetectRenames(true)
+
+        revWalk.markStart(head)
+        var commit: RevCommit? = revWalk.next()  // Move the walker to the head.
+        while (commit != null && commit != tail) {
+            val parentCommit: RevCommit? = revWalk.next()
+
+            // Smart casts are not yet supported for a mutable variable captured
+            // in an inline lambda, see
+            // https://youtrack.jetbrains.com/issue/KT-7186.
+            if (Logger.isDebug) {
+                val commitName = commit.getName()
+                val commitMsg = commit.getShortMessage()
+                Logger.debug { "commit: $commitName; '$commitMsg'" }
+                if (parentCommit != null) {
+                    val parentCommitName = parentCommit.getName()
+                    val parentCommitMsg = parentCommit.getShortMessage()
+                    Logger.debug {
+                        "parent commit: ${parentCommitName}; '${parentCommitMsg}'"
                     }
-                    // Commits are combined in pairs, an empty commit concatenated
-                    // to calculate the diff of the initial commit.
-                    subscriber.onNext(Commit())
-                } catch (e: Exception) {
-                    Logger.error(e, "Commit producing error")
-                    subscriber.onError(e)
                 }
-                subscriber.onComplete()
-            }  // TODO(anatoly): Rewrite diff calculation in non-weird way.
-            .pairWithNext()  // Pair commits to get diff.
-            .map { (new, old) ->
+                else {
+                    Logger.debug { "parent commit: null" }
+                }
+            }
+
+            val diffEntries = df.scan(parentCommit, commit);
+            val diffEdits = diffEntries.map { diff ->
+                Pair(diff, df.toFileHeader(diff).toEditList())
+            }
+            subscriber.onNext(Pair(commit, diffEdits))
+            commit = parentCommit
+        }
+
+        subscriber.onComplete()
+    }
+
+    fun getObservable(git: Git,
+                      jgitObservable: Observable<Pair<RevCommit, List<Pair<DiffEntry, EditList>>>>,
+                      repo: Repo,
+                      numCommits: Int = 0): Observable<Commit> {
+        var curNumCommits = 0
+
+        return Observable.create<Commit> { subscriber ->
+            jgitObservable.subscribe( { (jgitCommit, jgitDiffs) ->
```

Comment:
return jgitObsevable.map( { (jgitCommit, jgitDiffs) ->

Reply:
is that instead of Observable<Commit> object? What is return type would be in this case?